File: linux/data/AnalysisModels.cpp
#include "AnalysisModels.h"
#include "ApiClient.h"
#include "AnalysisChat.h"
#include "AnalysisRepository.h"
#include <QJsonDocument>
#include <QJsonArray>
#include <QJsonObject>
#include <QRegularExpression>
#include <QDebug>
#include <algorithm>

// ---------------------------------------------------------------------------
// Singleton
// ---------------------------------------------------------------------------

AnalysisModels& AnalysisModels::instance() {
    static AnalysisModels s_instance;
    return s_instance;
}

AnalysisModels::AnalysisModels(QObject *parent)
    : QObject(parent)
{
}

// ---------------------------------------------------------------------------
// fetchModels - dispatch by apiFormat
// ---------------------------------------------------------------------------

void AnalysisModels::fetchModels(const AppService &service, const QString &apiKey,
                                  const QString &customUrl,
                                  std::function<void(QStringList)> callback) {
    // Providers with hardcoded models and no model list API endpoint
    if (service.hardcodedModels.has_value() && !service.modelsPath.has_value()) {
        if (callback)
            callback(service.hardcodedModels.value());
        return;
    }

    switch (service.apiFormat) {
    case ApiFormat::Google:
        fetchGeminiModels(service, apiKey, callback);
        break;
    case ApiFormat::Anthropic:
        fetchClaudeModels(service, apiKey, callback);
        break;
    case ApiFormat::OpenAiCompatible:
        fetchOpenAiCompatibleModels(service, apiKey, customUrl, callback);
        break;
    }
}

// ---------------------------------------------------------------------------
// fetchOpenAiCompatibleModels
// ---------------------------------------------------------------------------

void AnalysisModels::fetchOpenAiCompatibleModels(const AppService &service, const QString &apiKey,
                                                   const QString &customUrl,
                                                   std::function<void(QStringList)> callback) {
    // Return hardcoded models if no models API path is set
    if (!service.modelsPath.has_value()) {
        if (callback)
            callback(service.hardcodedModels.value_or(QStringList()));
        return;
    }

    const QString url = ApiClient::modelsUrl(service, customUrl);
    if (url.isEmpty()) {
        if (callback)
            callback(service.hardcodedModels.value_or(QStringList()));
        return;
    }

    const bool isArrayFormat = (service.modelListFormat == QStringLiteral("array"));
    // Capture service by value for the lambda (id needed for filter)
    const AppService svc = service;

    ApiClient::instance().get(url, service, apiKey,
        [this, callback, isArrayFormat, svc](const QJsonObject &json, int statusCode) {
            Q_UNUSED(statusCode)

            QStringList modelIds;

            if (isArrayFormat) {
                // Some providers return a raw JSON array, which ApiClient wraps in {"data": [...]}
                const QJsonArray arr = json[QStringLiteral("data")].toArray();
                for (const auto &v : arr) {
                    const QString id = v.toObject()[QStringLiteral("id")].toString();
                    if (!id.isEmpty())
                        modelIds.append(id);
                }
            } else {
                // Standard OpenAI format: { "data": [ { "id": "..." }, ... ] }
                const QJsonArray arr = json[QStringLiteral("data")].toArray();
                for (const auto &v : arr) {
                    const QString id = v.toObject()[QStringLiteral("id")].toString();
                    if (!id.isEmpty())
                        modelIds.append(id);
                }
            }

            QStringList filtered = filterModels(modelIds, svc);
            if (callback)
                callback(filtered);
        });
}

// ---------------------------------------------------------------------------
// fetchClaudeModels
// ---------------------------------------------------------------------------

void AnalysisModels::fetchClaudeModels(const AppService &service, const QString &apiKey,
                                         std::function<void(QStringList)> callback) {
    const QString base = service.baseUrl.endsWith(QLatin1Char('/'))
                             ? service.baseUrl
                             : service.baseUrl + QLatin1Char('/');
    const QString url = base + QStringLiteral("v1/models");

    ApiClient::instance().get(url, service, apiKey,
        [callback](const QJsonObject &json, int statusCode) {
            Q_UNUSED(statusCode)

            QStringList modelIds;
            const QJsonArray arr = json[QStringLiteral("data")].toArray();
            for (const auto &v : arr) {
                const QString id = v.toObject()[QStringLiteral("id")].toString();
                if (!id.isEmpty() && id.startsWith(QStringLiteral("claude")))
                    modelIds.append(id);
            }

            modelIds.sort();
            if (callback)
                callback(modelIds);
        });
}

// ---------------------------------------------------------------------------
// fetchGeminiModels
// ---------------------------------------------------------------------------

void AnalysisModels::fetchGeminiModels(const AppService &service, const QString &apiKey,
                                         std::function<void(QStringList)> callback) {
    const QString base = service.baseUrl.endsWith(QLatin1Char('/'))
                             ? service.baseUrl
                             : service.baseUrl + QLatin1Char('/');
    const QString url = QStringLiteral("%1v1beta/models?key=%2").arg(base, apiKey);

    // Google auth uses query param, so create a temp service to avoid adding Bearer header
    AppService tempService;
    tempService.id = QStringLiteral("gemini_temp");
    tempService.apiFormat = ApiFormat::Google;

    ApiClient::instance().get(url, tempService, apiKey,
        [callback](const QJsonObject &json, int statusCode) {
            Q_UNUSED(statusCode)

            QStringList modelIds;
            const QJsonArray modelsArr = json[QStringLiteral("models")].toArray();
            for (const auto &v : modelsArr) {
                const QJsonObject modelObj = v.toObject();

                // Filter to models that support generateContent
                const QJsonArray methods = modelObj[QStringLiteral("supportedGenerationMethods")].toArray();
                bool supportsGenerate = false;
                for (const auto &m : methods) {
                    if (m.toString() == QStringLiteral("generateContent")) {
                        supportsGenerate = true;
                        break;
                    }
                }
                if (!supportsGenerate)
                    continue;

                // Strip "models/" prefix from the name field
                QString name = modelObj[QStringLiteral("name")].toString();
                if (name.startsWith(QStringLiteral("models/")))
                    name = name.mid(7); // len("models/") == 7
                if (!name.isEmpty())
                    modelIds.append(name);
            }

            modelIds.sort();
            if (callback)
                callback(modelIds);
        });
}

// ---------------------------------------------------------------------------
// filterModels - apply provider-specific regex filter and sort
// ---------------------------------------------------------------------------

QStringList AnalysisModels::filterModels(const QStringList &models, const AppService &service) {
    const QRegularExpression *regex = service.modelFilterRegex();

    QStringList result;
    if (regex) {
        for (const auto &model : models) {
            if (regex->match(model).hasMatch())
                result.append(model);
        }
    } else {
        result = models;
    }

    result.sort();
    return result;
}

// ---------------------------------------------------------------------------
// testModel - send test prompt and return error (empty = success)
// ---------------------------------------------------------------------------

void AnalysisModels::testModel(const AppService &service, const QString &apiKey,
                                const QString &model,
                                std::function<void(QString error)> callback) {
    const QString prompt = AnalysisRepository::testPrompt;

    QList<ChatMessage> messages;
    messages.append(ChatMessage(QStringLiteral("user"), prompt));

    ChatParameters params;

    AnalysisChat::instance().sendChatMessage(service, apiKey, model, messages, params, QString(),
        [callback](const QString &response, const QString &error) {
            Q_UNUSED(response)
            if (callback)
                callback(error);
        });
}

// ---------------------------------------------------------------------------
// testModelWithPrompt - send custom prompt and return response + error
// ---------------------------------------------------------------------------

void AnalysisModels::testModelWithPrompt(const AppService &service, const QString &apiKey,
                                           const QString &model, const QString &prompt,
                                           std::function<void(QString response, QString error)> callback) {
    QList<ChatMessage> messages;
    messages.append(ChatMessage(QStringLiteral("user"), prompt));

    ChatParameters params;

    AnalysisChat::instance().sendChatMessage(service, apiKey, model, messages, params, QString(),
        [callback](const QString &response, const QString &error) {
            if (callback)
                callback(response, error);
        });
}

File: iOS/ai/Data/AnalysisModels.swift
import Foundation

// MARK: - Model Fetching + API Testing

extension AnalysisRepository {

    // MARK: - Dispatch

    func fetchModels(service: AppService, apiKey: String) async throws -> [String] {
        switch service.apiFormat {
        case .anthropic:
            return try await fetchClaudeModels(service: service, apiKey: apiKey)
        case .google:
            return try await fetchGeminiModels(service: service, apiKey: apiKey)
        case .openaiCompatible:
            return try await fetchModelsOpenAiCompatible(service: service, apiKey: apiKey)
        }
    }

    // MARK: - OpenAI Compatible

    private func fetchModelsOpenAiCompatible(service: AppService, apiKey: String) async throws -> [String] {
        // Handle hardcoded models (no API endpoint)
        if let hardcoded = service.hardcodedModels, service.modelsPath == nil {
            return hardcoded
        }

        let url = ApiClient.modelsUrl(for: service)
        guard !url.isEmpty else {
            return service.hardcodedModels ?? []
        }

        let headers = ApiClient.openAiHeaders(apiKey: apiKey)

        var modelIds: [String]

        if service.modelListFormat == "array" {
            // Some providers return a plain array instead of {data: [...]}
            let (models, _, _): ([OpenAiModel], Int, [String: String]) =
                try await ApiClient.shared.sendRequest(url: url, method: "GET", headers: headers, responseType: [OpenAiModel].self)
            modelIds = models.compactMap(\.id)
        } else {
            let (response, _, _): (OpenAiModelsResponse, Int, [String: String]) =
                try await ApiClient.shared.sendRequest(url: url, method: "GET", headers: headers, responseType: OpenAiModelsResponse.self)
            modelIds = response.data?.compactMap(\.id) ?? []
        }

        // Apply provider-specific filter
        let filtered = service.modelFilter != nil ? modelIds.filter { service.matchesFilter($0) } : modelIds

        return filtered.sorted()
    }

    // MARK: - Claude Models

    private func fetchClaudeModels(service: AppService, apiKey: String) async throws -> [String] {
        let normalizedBase = ApiClient.normalizeBaseUrl(service.baseUrl)
        let url = "\(normalizedBase)v1/models"
        let headers = ApiClient.anthropicHeaders(apiKey: apiKey)

        let (response, _, _): (ClaudeModelsResponse, Int, [String: String]) =
            try await ApiClient.shared.sendRequest(url: url, method: "GET", headers: headers, responseType: ClaudeModelsResponse.self)

        let models = (response.data ?? []).compactMap(\.id).filter { $0.hasPrefix("claude") }
        return models.sorted()
    }

    // MARK: - Gemini Models

    private func fetchGeminiModels(service: AppService, apiKey: String) async throws -> [String] {
        let url = ApiClient.geminiModelsUrl(baseUrl: service.baseUrl, apiKey: apiKey)
        let headers = ApiClient.geminiHeaders()

        let (response, _, _): (GeminiModelsResponse, Int, [String: String]) =
            try await ApiClient.shared.sendRequest(url: url, method: "GET", headers: headers, responseType: GeminiModelsResponse.self)

        let models = response.models?
            .filter { $0.supportedGenerationMethods?.contains("generateContent") ?? false }
            .compactMap { $0.name?.replacingOccurrences(of: "models/", with: "") }
            ?? []

        return models.sorted()
    }

    // MARK: - Test Model

    func testModel(service: AppService, apiKey: String, model: String) async -> String? {
        let response = await analyze(service: service, apiKey: apiKey, prompt: Self.testPrompt, model: model)
        return response.isSuccess ? nil : (response.error ?? "Unknown error")
    }

    func testModelWithPrompt(service: AppService, apiKey: String, model: String, prompt: String) async -> (String?, String?) {
        let response = await analyze(service: service, apiKey: apiKey, prompt: prompt, model: model)
        if response.isSuccess {
            return (response.analysis, nil)
        } else {
            return (nil, response.error ?? "Unknown error")
        }
    }

    // MARK: - Test API with Raw JSON

    func testApiConnectionWithJson(
        service: AppService,
        apiKey: String,
        jsonBody: String,
        customUrl: String? = nil
    ) async -> (String?, String?, Int?) {
        // Force stream: false for testing
        var modifiedJson = jsonBody
        if let data = jsonBody.data(using: .utf8),
           var json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            json["stream"] = false
            if let modified = try? JSONSerialization.data(withJSONObject: json),
               let str = String(data: modified, encoding: .utf8) {
                modifiedJson = str
            }
        }

        let url: String
        if let custom = customUrl, !custom.isEmpty {
            url = custom
        } else {
            url = ApiClient.chatUrl(for: service)
        }

        let headers = ApiClient.headersForService(service, apiKey: apiKey)

        // For Gemini, append key to URL
        var finalUrl = url
        if service.apiFormat == .google, !finalUrl.contains("key=") {
            finalUrl += (finalUrl.contains("?") ? "&" : "?") + "key=\(apiKey)"
        }

        guard let body = modifiedJson.data(using: .utf8) else {
            return (nil, "Invalid JSON", nil)
        }

        do {
            let (data, statusCode, _) = try await ApiClient.shared.sendRawRequest(url: finalUrl, headers: headers, body: body)
            let responseBody = String(data: data, encoding: .utf8) ?? "No response body"

            if (200...299).contains(statusCode) {
                return (responseBody, nil, statusCode)
            } else {
                return (nil, "HTTP \(statusCode): \(responseBody)", statusCode)
            }
        } catch {
            return (nil, error.localizedDescription, nil)
        }
    }
}
